cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp

```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
             std::int32_t pos = (std::int32_t)(rnd * (partition_end - partition_start - 1));
             pos = pos < 0 ? 0 : pos;
             const Float pivot = values[partition_start + pos];
-            std::int32_t split_index = row_partitioning_kernel(item,
-                                                               values,
-                                                               indices,
-                                                               partition_start,
-                                                               partition_end,
-                                                               pivot);
-
-            if ((split_index) == k)
+            auto partition_size = partition_end - partition_start - 1;
+            if (partition_size > local_size) {
+                std::int32_t split_index = row_partitioning_kernel(item,
+                                                                   values,
+                                                                   indices,
+                                                                   partition_start,
+                                                                   partition_end,
+                                                                   pivot);
+                if ((split_index) == k)
+                    break;
+                if (split_index > k)
+                    partition_end = split_index;
+                if (split_index < k)
+                    partition_start = split_index;
+            }
+            else {
+                auto rem_k = k - partition_start;
+                if (rem_k <= 0)
+                    break;
```

Comment:
Why this exit condition is valid for this branch only?

Reply:
Removed